lib/Cmt.ml

```diff
@@ -88,25 +88,115 @@ end
 
 type pos = Before | Within | After
 
-let unindent_lines ~offset first_line tl_lines =
+type decoded_kind =
+  | Verbatim of string
+  | Doc of string
+  | Normal of string
+  | Code of string
+  | Asterisk_prefixed of string list
+
+type decoded = {prefix: string; suffix: string; kind: decoded_kind}
+
+(** [opn_offset] indicates at which column the body of the comment starts. *)
+let unindent_lines ~opn_offset first_line tl_lines =
   let indent_of_line s =
     (* index of first non-whitespace is indentation, None means white line *)
     String.lfindi s ~f:(fun _ c -> not (Char.is_whitespace c))
   in
   (* The indentation of the first line must account for the location of the
-     comment opening *)
-  let fl_spaces = Option.value ~default:0 (indent_of_line first_line) in
-  let fl_indent = fl_spaces + offset in
+     comment opening. Don't account for the first line if it's empty. *)
+  let fl_spaces, fl_indent =
+    match indent_of_line first_line with
+    | Some i -> (i, i + opn_offset)
+    | None -> (String.length first_line, Stdlib.max_int)
+  in
   let min_indent =
     List.fold_left ~init:fl_indent
       ~f:(fun acc s ->
-        Option.value_map ~default:acc ~f:(min acc) (indent_of_line s) )
+        match indent_of_line s with Some i -> min acc i | None -> acc )
       tl_lines
   in
   (* Completely trim the first line *)
   String.drop_prefix first_line fl_spaces
   :: List.map ~f:(fun s -> String.drop_prefix s min_indent) tl_lines
 
-let unindent_lines ~offset = function
+let unindent_lines ~opn_offset txt =
+  match String.split ~on:'\n' txt with
   | [] -> []
-  | hd :: tl -> unindent_lines ~offset hd tl
+  | hd :: tl -> unindent_lines ~opn_offset hd tl
+
+let split_asterisk_prefixed =
+  let line_is_asterisk_prefixed s = String.is_prefix s ~prefix:"*" in
+  let line_is_asterisk_or_space_prefixed s =
+    if String.is_empty s then true
+    else match s.[0] with '*' | ' ' -> true | _ -> false
+  in
+  (* Whether every lines starts with "*" or " ". At least one line must start
+     with a "*" and completely empty lines are allowed. *)
```

Comment:
This creates regressions such as:
```diff
                 (*with exn ->
-                    failwiths "binary_search bug"
-                    (exn, `length length, `search_key search_key, `pos pos, `len len)
-                    <:sexp_of< exn * [ `length of int ] * [ `search_key of int ]
+                 *  failwiths "binary_search bug"
+                 *  (exn, `length length, `search_key search_key, `pos pos, `len len)
+                 *  <:sexp_of< exn * [ `length of int ] * [ `search_key of int ]
                  * [ `pos of int ] * [ `len of int ] >>*)
```

Reply:
Good catch, I'll make this parsing stricter.